File: fib-memorization.clj
;; declare 함수 - 바인딩되지 않은 심벌 생성, 
;; 해당 함수가 정의되어 바인딩 되기 전 다른 함수에서 사용 가능
(declare fib)

(defn fib-w [n]
    (cond 
        (< n 1) nil
        (<= n 2) 1
        :else (+ (fib (dec n)) (fib (- n 2)))
    )
)

;; memoize - 함수인 인수 f를 받아 새로운 함수 g를 반환
(def fib (memoize fib-w))

File: fib-memorization.js
function fib_w(n) {
    if (n < 1) {
        return null;
    } else if (n <= 2) {
        return 1;
    } else {
        return fib(n - 1) + fib(n - 2);
    }
}

function memoize(fn) {
    const cache = new Map();

    return function (n) {
        if (cache.has(n)) {
            return cache.get(n);
        }

        const result = fn(n);
        cache.set(n, result);
        return result;
    }
}

const fib = memoize(fib_w);

console.log(fib(3));
console.log(fib(5));
console.log(fib(11));
console.log(fib(50));